all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+of which are something users should be required to think about. This proposal aims to reduce the cognitive load of
+configuring workers by automatically tuning various worker internals at runtime without (or with configurable) user
+intervention.
+
+## Areas of configuration
+There are a few different configuration areas that we can autotune:
+
+### Concurrent Pollers
+Right now the number of pollers used for polling workflow and activity tasks is fixed to a maximum defined by the user.
+In Core: `max_concurrent_wft_polls` and `max_concurrent_at_polls`. There's no need for this number to be fixed, and we
+can and should scale it up and down as demand dictates.  This work is largely already designed and implemented in Go:
+[see the branch
+here](https://github.com/temporalio/sdk-go/compare/master...Quinn-With-Two-Ns:sdk-go:poller-autottune-v0.2). Hence this
+doc will not focus on this area in detail.
+
+It is important to note that there is a distinction between the active number of poller *routines* and the active number
+of *open actual RPC calls*. The number of routines may be scaled independently of the number of open RPC calls at any
+time. The former may have a floor, whereas the latter must always be able to go to 0. Users are unlikely to care about
+the number of routines as they are relatively cheap. They *do*, however, care about how quickly we will change the
+number of active RPCs. The floor on the number of routines effectively determines how quickly we are willing to consume
+available slots and thus make actual RPC calls. IE: If the floor on the number of routines is 10, and 10 slots open up
+at the same time, we will immediately make 10 RPC calls. 
+
+In Go the name for these options is already misleading: `MaxConcurrentWorkflowTaskPollers` is not accurate, it's really
+`NumWorkflowTaskPollers` (or routines), what it probably should be is `MaxConcurrentWorkflowTaskPolls`, since the number
+of open RPC calls is actually what's fluctuating.
+
+
+### Slot tuning & Cache Size
+
+Workers have a configurable number of slots for task types. In Core: `max_outstanding_workflow_tasks`,
+`max_outstanding_activities`, and `max_outstanding_local_activities`. These are set by the user and can't change at
+runtime. Autotuning of these values seeks to allow them to change at runtime according to available resources on
+the worker. It is not feasible for us to choose *optimal* values for these at runtime, because we cannot know a-priori
+about how much resources any given workflow or activity task will consume - only the user can (maybe) know this.
+However, we can provide reasonable default autotuning implementations while also allowing users to provide their own.
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
```

Comment:
Somewhere before this point, I feel that I'm missing an explicit statement of what the problem to be solved is (expressed in terms of the technical concepts introduced in the paragraphs above, but "slot management" needs some background explanation).

Reply:
my earlier comment seems to be a long-winded expression of what Dan is saying here.